oss_src/sframe_query_engine/algorithm/ec_permute.cpp

```diff
@@ -0,0 +1,583 @@
+  for (size_t i = 0;i < out_column_types.size(); ++i) {
+    if (indirect_column[i]) out_column_types[i] = flex_type_enum::INTEGER;
+  }
+  output.open_for_write(input.column_names(), out_column_types, "", num_buckets);
+  auto writer = output.get_internal_writer();
+
+  // prepare all the readers
+  std::vector<std::shared_ptr<sarray<flexible_type>::reader_type>> readers(input.num_columns());
+  for (size_t i = 0;i < input.num_columns(); ++i) {
+    auto cur_column = input.select_column(i);
+    readers[i] = std::shared_ptr<sarray<flexible_type>::reader_type>(cur_column->get_reader());
+  }
+  // now. the challenge here is that the natural order of the sframe is not 
+  // necessarily good for forward map lookups. The forward map lookup has to
+  // really fast. Instead we are going to do it this way:
+  //  - Use the SFRAME_SORT_BUFFER_SIZE to estimate how much forward map 
+  //  we can keep in memory at any one time. Then in parallel over 
+  //  columns of the sframe.
+  size_t max_forward_map_in_memory = SFRAME_SORT_BUFFER_SIZE / sizeof(flexible_type);
+  auto forward_map_reader = forward_map->get_reader();
+  std::vector<flexible_type> forward_map_buffer;
+  logstream(LOG_INFO) << "Beginning Scatter"  << std::endl;
+  logstream(LOG_INFO) << "Maximum forward map in memory " 
+                      << max_forward_map_in_memory 
+                      << std::endl;
+
+  for (size_t forward_map_start = 0; 
+       forward_map_start < forward_map->size(); 
+       forward_map_start += max_forward_map_in_memory) {
+    size_t forward_map_end = std::min(forward_map_start + max_forward_map_in_memory, 
+                                      forward_map->size());
+    logstream(LOG_INFO) << "Processing rows " 
+                        << forward_map_start 
+                        << " to " << forward_map_end << std::endl;
+    forward_map_reader->read_rows(forward_map_start, forward_map_end, 
+                                  forward_map_buffer);
+
+    // now in parallel over columns.
+    atomic<size_t> col_number = 0;
+    in_parallel(
+        [&](size_t unused, size_t unused2) {
+          while(1) {
+            // this is the column we are now processing
+            size_t column_id = col_number.inc_ret_last();
+            if (column_id >= input.num_columns()) return;
+
+            if (indirect_column[column_id]) {
+              // if this is to be an indirect column, 
+              // we use write the row number as the value
+              for (size_t actual_row = forward_map_start; actual_row < forward_map_end; ++actual_row) {
+                size_t output_row = 
+                    forward_map_buffer[actual_row - forward_map_start].get<flex_int>();
+                size_t output_segment = output_row / rows_per_bucket;
+                if (output_segment >= num_buckets) output_segment = num_buckets - 1;
+                writer->write_segment(column_id, output_segment, flexible_type(actual_row));
```

Comment:
Technically no. Its a possible optimization. Though probably not too significant an optimization since the row numbers are small and the indirect columns are  big.